hilti/toolchain/include/ast/declaration.h

```diff
@@ -65,16 +71,30 @@ inline bool operator==(const Declaration& x, const Declaration& y) {
 
 inline bool operator!=(const Declaration& d1, const Declaration& d2) { return ! (d1 == d2); }
 
-} // namespace detail
-} // namespace declaration
-
-using Declaration = declaration::detail::Declaration;
-using declaration::detail::to_node;
-
+namespace declaration {
 /** Constructs an AST node from any class implementing the `Declaration` interface. */
 template<typename T, typename std::enable_if_t<std::is_base_of<trait::isDeclaration, T>::value>* = nullptr>
 inline Node to_node(T t) {
     return Node(Declaration(std::move(t)));
 }
+} // namespace declaration
+
+/**
+ * Base class for classes implementing the `Declaration` interface. This class
+ * provides implementations for some interface methods shared that are shared
+ * by all declarations.
+ */
+class DeclarationBase : public NodeBase, public hilti::trait::isDeclaration {
+public:
+    using NodeBase::NodeBase;
+
+    /** Implements the `Declaration` interface. */
+    const ID& canonicalID() const { return _id; }
+    /** Implements the `Declaration` interface. */
+    void setCanonicalID(ID id) { _id = std::move(id); }
```

Comment:
This is a dangerous interface as it makes it very easy to accidentally create canonical ID which are empty via a the default ctr; in fact I am seeing a number of cases where `canonicalID` returns an empty string which is not at all unique.

Should we maybe instead change `canonicalID` and `_id` to be `optional<ID>`s so users have at least a chance to detect completely unset canonical IDs? That still wouldn't solve the issue that the way these IDs are constructed makes it hard to ensure uniqueness though.

Reply:
I'm adding asserts, as discussed.